### ISSUE TITLE
flatten() is not ONNX exportable with dynamic axes

### ISSUE THREAD

**Issue by pfeatherstone** [pytorch/pytorch#42993]
## Expected behavior

I would expect the following to not break

```python
import onnxruntime
import torch
import torch.nn as nn

class dummyNet(nn.Module):
    def __init__(self):
        super(dummyNet, self).__init__()
        self.n =  nn.Conv2d(1, 35, 1)

    def forward(self, x):
        x = self.n(x)
        x = x.unsqueeze(1)
        x = x.permute(0, 1, 3, 2, 4).contiguous()
        x = x.flatten(3, 4)
        return x

n = dummyNet()
x = torch.randn(1,1,1024,2)
ort_x = torch.randn(1,1,4096,2).numpy()

torch.onnx.export(n,
                             (x,),
                             "d.onnx",
                             opset_version=12,
                             input_names=['in'],
                             output_names=['out'],
                             dynamic_axes={'in': [0, 2], 'out': [0,2]})

ort_session = onnxruntime.InferenceSession("d.onnx")
ort_inputs = {}
ort_inputs[ort_session.get_inputs()[0].name] = ort_x
ort_outs = ort_session.run(None, ort_inputs)
```

If add the following lines
```
ort_inputs = ort_session.get_inputs()
print(ort_inputs[0].shape)
```
I get
```
['in_dynamic_axes_1', 1, 'in_dynamic_axes_2', 2]
```

So it thinks the batch dimension and height dimension are dynamic.
But if you run the code you get stack trace:
```
2020-08-13 07:13:26.266184390 [E:onnxruntime:, sequential_executor.cc:309 Execute] Non-zero status code returned while running Reshape node. Name:'Reshape_4' Status Message: /onnxruntime_src/onnxruntime/core/providers/cpu/tensor/reshape_helper.h:43 onnxruntime::ReshapeHelper::ReshapeHelper(const onnxruntime::TensorShape&, std::vector<long int>&) gsl::narrow_cast<int64_t>(input_shape.Size()) == size was false. The input tensor cannot be reshaped to the requested shape. Input shape:{1,1,4096,35,2}, requested shape:{1,1,1024,70}
```

So yes the error occurs in ONNXRUNTIME, but it is only doing what it is told, and the pytorch ONNX exporter has told it to flatten {1,1,4096,35,2} to {1,1,1024,70}

## Environment

Collecting environment information...
PyTorch version: 1.6.0
Is debug build: No
CUDA used to build PyTorch: 10.2
OS: Ubuntu 18.04.4 LTS
GCC version: (Ubuntu 8.4.0-1ubuntu1~18.04) 8.4.0
Clang version: 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
CMake version: version 3.16.6
Python version: 3.6 (64-bit runtime)
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: GeForce RTX 2070 SUPER
GPU 1: GeForce RTX 2070 SUPER

Nvidia driver version: 440.33.01
cuDNN version: /usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn.so.7.6.5
Versions of relevant libraries:
[pip3] efficientnet-pytorch==0.6.3
[pip3] msgpack-numpy==0.4.5
[pip3] numpy==1.19.1
[pip3] pytorch-metric-learning==0.9.85
[pip3] segmentation-models-pytorch==0.1.0
[pip3] torch==1.6.0
[pip3] torch-optimizer==0.0.1a10
[pip3] torchaudio==0.6.0
[pip3] torchvision==0.7.0
[conda] Could not collect

cc @houseroad @spandantiwari @lara-hdr @BowenBao @neginraoof

**Comment by pfeatherstone** [comment]
In fact, the pytorch ONNX exporter has exported the flatten layer to a "Reshape node". Is that correct?

**Comment by pfeatherstone** [comment]
This issue https://github.com/microsoft/onnxruntime/issues/1088 says that Flatten is supported in ONNXRUNTIME. So I don't know why pytorch is not using it.

**Comment by garymm** [comment]
The provided code works with onnxruntime 1.11.0 and torch 1.10.0. Please re-open with version info if this is still a problem.
